src/field_theory/krull_topology.lean

```diff
@@ -287,3 +289,287 @@ begin
+  rw ultrafilter.glued_generators_of_pushforwards_function_spec' h_int f hE ⟨x, hxE⟩ at hxy,
+  rw ultrafilter.glued_generators_of_pushforwards_function_spec' h_int f hE ⟨y, hyE⟩ at hxy,
+  simpa using ring_hom.injective (f.generator_of_pushforward hE).to_ring_hom hxy,
+end
+
+lemma eq_of_map_le {f : L →ₐ[K] L} (h_findim : finite_dimensional K E) (h_map_le : E.map f ≤ E) :
+  E.map f = E :=
+begin
+  have hf_inj : function.injective f := ring_hom.injective f.to_ring_hom,
+  haveI hE_submod_fin : finite_dimensional K E.to_subalgebra.to_submodule,
+  { exact h_findim },
+  have h_finrank_eq : finite_dimensional.finrank K (E.map f) =
+    finite_dimensional.finrank K E,
+  { exact (linear_equiv.finrank_eq (submodule.equiv_map_of_injective (f.to_linear_map)
+    hf_inj E.to_subalgebra.to_submodule)).symm },
+  exact intermediate_field.to_subalgebra_eq_iff.mp (subalgebra.to_submodule_injective
+  (finite_dimensional.eq_of_le_of_finrank_eq h_map_le h_finrank_eq)),
+end
+
+lemma ultrafilter.glued_generators_of_pushforwards_alg_hom_surjective (h_int :
+  algebra.is_integral K L) (f : ultrafilter (L →ₐ[K] L)) :
+  function.surjective (ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int) :=
+begin
+  intro y,
+  let p := minpoly K y,
+  let S := (p.map (algebra_map K L)).roots.to_finset,
+  let E := intermediate_field.adjoin K (S : set L),
+  have hE_findim : finite_dimensional K E :=
+  intermediate_field.adjoin.finite_dimensional_of_finite_set  (finset.finite_to_set S)
+  (λ x hx, h_int x),
+  let σ := ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int,
+  have hσSS : σ '' S ⊆ S,
+  { rintros x ⟨a, ha, hax⟩,
+    rw ← hax,
+    simp only [finset.mem_coe, multiset.mem_to_finset],
+    rw polynomial.mem_roots,
+    { rw [polynomial.is_root.def, ← polynomial.eval₂_eq_eval_map,
+        ← polynomial.alg_hom_eval₂_algebra_map, (show 0 = σ 0, by simp)],
+      apply congr_arg σ,
+      rwa [finset.mem_coe, multiset.mem_to_finset, polynomial.mem_roots, polynomial.is_root.def,
+        ← polynomial.eval₂_eq_eval_map] at ha,
+      exact polynomial.map_monic_ne_zero (minpoly.monic (h_int y)) },
+    exact polynomial.map_monic_ne_zero (minpoly.monic (h_int y)) },
+  have hSE : (S : set L) ⊆ E := intermediate_field.gc.le_u_l (S : set L),
+  have hσSE : σ '' S ⊆ E := set.subset.trans hσSS hSE,
+  have h1 : E.map σ = intermediate_field.adjoin K (σ '' S) := intermediate_field.adjoin_map K S σ,
+  have h2 : intermediate_field.adjoin K (σ '' S) ≤ E := intermediate_field.gc.l_le hσSE,
+  rw ← h1 at h2,
+  have h3 := eq_of_map_le hE_findim h2,
+  have hyE : y ∈ E,
+  { have hyS : y ∈ S,
+    { rw [multiset.mem_to_finset, polynomial.mem_roots],
+      { rw [polynomial.is_root.def, ← polynomial.eval₂_eq_eval_map, ← polynomial.aeval_def],
+        exact minpoly.aeval K y },
+      { exact polynomial.map_monic_ne_zero (minpoly.monic (h_int y)) } },
+    exact hSE hyS },
+  rw ← h3 at hyE,
+  rcases hyE with ⟨a, ha, hay⟩,
```

Comment:
Golfed (using `root_set` and focusing on `S` rather than `E`).
```suggestion
  let S := p.root_set L,
  let σ := ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int,
  have hp : p ≠ 0 := (minpoly.monic (h_int y)).ne_zero,
  have hσSS : σ '' S ⊆ S,
  { rintros x ⟨a, ha, rfl⟩,
    rw polynomial.mem_root_set hp at ha,
    rw [polynomial.mem_root_set hp, polynomial.aeval_alg_hom_apply, ha, σ.map_zero] },
  have h3 : σ '' S = S :=
  eq_of_subset_of_card_le hσSS (ge_of_eq (card_image_of_injective S σ.to_ring_hom.injective)),
  have hy : y ∈ σ '' S,
  { rw [h3, polynomial.mem_root_set hp, minpoly.aeval] },
  rcases hy with ⟨a, ha, hay⟩,
```